src/weights.jl

```diff
@@ -2,43 +2,112 @@
         sum::S
+        bias::Real
     end
 else
-    immutable WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
+    abstract AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
+
+    immutable Weights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+        values::V
+        sum::S
+        bias::Real
+    end
+
+    immutable FrequencyWeights{S<:Integer, T<:Integer, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
         values::V
         sum::S
+        bias::Int
+    end
+
+    immutable ProbabilityWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+        values::V
+        sum::S
+        bias::Real
     end
 end
 
 """
-    WeightVec(vs, [wsum])
+    Weights(vs, [wsum])
 
-Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
+Construct a `Weights` with weight values `vs` and sum of weights `wsum`.
 If omitted, `wsum` is computed.
 """
-function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
-    return WeightVec{S, eltype(vs), V}(vs, s)
+function Weights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs); corrected::Bool=true)
+    if isempty(vs) || !corrected
+        return Weights{typeof(s), eltype(vs), V}(vs, s, s)
+    else
+        return Weights{S, eltype(vs), V}(vs, s, s * (1 - sum(normalize(vs, 1) .^ 2)))
+    end
+end
+
+function FrequencyWeights{S<:Integer, V<:IntegerVector}(vs::V, s::S=sum(vs); corrected::Bool=true)
+    return FrequencyWeights{S, eltype(vs), V}(vs, s, s - Int(corrected))
 end
 
+# TODO: constructor for ProbabilityWeights, but I'm not familiar with how bias correction works with these
+# types of weights or if bias correction even makes sense.
+# https://en.wikipedia.org/wiki/Inverse_probability_weighting
+
 """
     weights(vs)
 
-Construct a `WeightVec` from a given array.
+Construct a `Weights` type from a given array.
+"""
+weights(vs::RealVector, corrected=true) = Weights(vs; corrected=corrected)
+weights(vs::RealArray, corrected=true) = Weights(vec(vs); corrected=corrected)
+
 """
-weights(vs::RealVector) = WeightVec(vs)
-weights(vs::RealArray) = WeightVec(vec(vs))
+    frequency(vs)
+
+Construct a `FrequencyWeights` type from a given array.
+"""
+frequency(vs::RealVector, corrected=true) = FrequencyWeights(vs; corrected=corrected)
+frequency(vs::RealArray, corrected=true) = FrequencyWeights(vec(vs); corrected=corrected)
+
+"""
+    exponential(n, [λ])
+
+Constructs a `Weights` type with a desired length `n` and smoothing factor `λ`,
+where each element is set to `λ * (1 - λ)^(1 - i)`.
+
+# Arguments
+* `n::Integer`: the desired length of the `Weights`
+* `λ::Real`: is a smoothing factor or rate paremeter between 0 .. 1.
+    As this value approaches 0 the resulting weights will be almost equal(),
+    while values closer to 1 will put higher weight on the end elements of the vector.
+"""
+function exponential(n::Integer, λ::Real=0.99)
+    @assert 0 <= λ <= 1 && n > 0
```

Comment:
Typically for argument checking it's better to use `condition || throw(ArgumentError("..."))`. In this case,
```julia
n > 0 || throw(ArgumentError("cannot construct weights of length < 1"))
0 <= λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
```
That provides more descriptive error messages for the user, plus it's better to use specifically-typed exceptions where possible.